torch/csrc/jit/script/compiler.cpp

```diff
@@ -72,14 +176,110 @@ struct to_ir {
       }
     }
   }
+
+  std::shared_ptr<Environment> emitSingleIfBranch(
+      Block* b,
+      const List<Stmt> branch,
+      std::unordered_set<std::string>* mutated_parent_values) {
+    environment_stack = std::make_shared<Environment>(b, environment_stack);
+    WithInsertPoint guard(*def.graph, b);
+    emitStatements(branch);
+
+    for (const auto& kv : environment_stack->value_table) {
+      if (environment_stack->findInParentFrame(kv.first)) {
+        mutated_parent_values->insert(kv.first);
+      }
+    }
+    auto save_env = environment_stack;
+    environment_stack = environment_stack->next;
+    return save_env;
+  }
+
   void emitIf(const If& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(kIf, 0));
+    n->addInput(cond_value);
+    auto* true_block = n->addBlock();
+    auto* false_block = n->addBlock();
+
+    // Emit both blocks once to get the union of all mutated values
+    std::unordered_set<std::string> mutated_parent_values;
+    auto save_true = emitSingleIfBranch(
+        true_block, stmt.trueBranch(), &mutated_parent_values);
+    auto save_false = emitSingleIfBranch(
+        false_block, stmt.falseBranch(), &mutated_parent_values);
+
+    std::vector<std::string> sorted_mutations(
+        mutated_parent_values.begin(), mutated_parent_values.end());
+    std::sort(sorted_mutations.begin(), sorted_mutations.end());
+
+    // Register outputs in each block
+    for (const auto& x : sorted_mutations) {
+      true_block->registerOutput(save_true->getVar(x, stmt));
+    }
+    for (const auto& x : sorted_mutations) {
+      false_block->registerOutput(save_false->getVar(x, stmt));
+    }
+
+    // Add op outputs
+    for (const auto& x : sorted_mutations) {
+      environment_stack->setVar(x, n->addOutput());
+    }
   }
 
   void emitWhile(const While& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    // Emits a loop operators conforming to the semantics specified at
+    // https://github.com/onnx/onnx/blob/master/docs/Operators.md#experimental-loop
+    // TODO: implement scan_outputs
+
+    Value* trip_count_dummy = emitConst(0, "i")[0];
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(kLoop, 0));
+    n->addInput(trip_count_dummy);
```

Comment:
the fact that loops take the trip count as an input is really weird. Is that what ONNX does?

Reply:
Let me rename this to `max_trip_count`. In the expression `for (size_t i=0; i < M && cond; ++i)` this would represent the value `M`

Reply:
But note this is a dummy value because this is a while loop